L69-cpp-core-event-manager.md

```diff
@@ -0,0 +1,282 @@
+  virtual void NotifyWhenWritableOrTimeout(std::function<void(NotificationOutcome)> upcall, gpr_timespec deadline) = 0;
+
+  // Clears the descriptor's status. This allows the NotifyWhen* calls to 
+  // distinguish between old and new events.
+  virtual void ClearReadable() = 0;
+  virtual void ClearWritable() = 0;
+  virtual void ClearHasError() = 0;
+
+  // Forces the descriptor’s status to change, and runs the closures if registered. 
+  // This is especially useful for tearing down closures associated with this 
+  // descriptor by forcing them to run without waiting for the event on the 
+  // underlying socket.
+  virtual void ForceReadable() = 0;
+  virtual void ForceWritable() = 0;
+  virtual void ForceHasError() = 0;
+
+  // Returns a human-readable description of the file descriptor state.
+  virtual string ToString() const = 0;
+};
+
+class EpollEventManagerInterface: public BaseEventManagerInterface {
+ public:
+  virtual ~EpollEventManagerInterface();
+
+  virtual EpollDescriptorInterface* RegisterFileDescriptor(int fd) = 0;
+
+  virtual void DeleteDescriptor(EpollDescriptorInterface* d) = 0;
+
+  // Adds a new one-shot alarm with a slack, which will be scheduled at the given 
+  // deadline plus the timespan specified in slack. Reducing the precision of 
+  // alarms through a timer slack allows the EventManager to perform 
+  // timer-coalescing, which minimizes thread wake-up and has the potential of 
+  // substantially reducing CPU utilization.
+  virtual void ScheduleAlarmInWithSlack(gpr_timespec deadline, std::function<void(bool)> c, Priority priority, TaskKey* key, gpr_timespec slack) = 0;
+};
+```
+
+Note that there will be no native implementation provided by gRPC for the epoll based interface in the first step. Instead, gRPC will provide a default event manager implementation based on libuv. The libuv EM implementation will be applied to all the gRPC languages based on gRPC core, and the users may choose to override it with their custom implementation of the EM interface. gRPC may release more native event manager implementations down the road map.
+```cpp
+// Represents a file descriptor when using libuv.
+class LibuvDescriptorInterface {
+ public:
+  virtual ~LibuvDescriptorInterface();
+
+  // Returns the underlying TCP or UDP handle. Valid until Close() is called
+  // (i.e. not thread-safe with Close()).
+  virtual uv_handle_t fd() const = 0;
+
+  // Retruns the EventManager object where this descriptor is registered.
+  virtual LibuvEventManagerInterface* event_manager() const = 0;
+
+  // Executes all the registered notification closures, and closes the
+  // underlying TCP or UDP handle.
+  //
+  // Note that we cannot release the underlying TCP or UDP handle, since it is
+  // initialized with the libuv event loop of LibuvEventManager, and it is not
+  // thread-safe to access the handle externally.
+  virtual void Close() = 0;
```

Comment:
Can either of these member functions be const?

Reply:
Can you elaborate? Close() cannot be const because it *closes* the thing, right?

Reply:
Hmmm... Not sure anymore myself. What I see now wrt. `const` looks fine here.